source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -1167,6 +1167,44 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         addUsersToWorkList(newStore);
     }
 
+    void processNonUniformResourceIndex(IRInst* nonUniformResourceIndexInst)
+    {
+        // implement the translation to spirv by waking up the use-def chain
+        // when we reach the leaf node (texture
+        // trace the use-def chain from nonuniformresource index def all the way to the
+        // textures / resources
+        // go through GEP to identify the pointer access
+        // go throuhg ExtractElement instruction for array access
+        // go through Load Inst to
+        // go through IntCasts to deal with u32 -> i32 / vice-versa
+        // add the decoration to the first Use that is either an extract element  / a load of
+        // an extract element address instruction
+        // if the first use is an extract element, find all the loads that uses the element,
+        // and the decorate them
+        //
+        // if the first use is a getelementptr, we may want to decorate the get element ptr itself,
+        // because user might
+        // call an image atomic operation, that operates on the ptr instead of the texture / resource
+        // todo rename the decoration itself nonuniformresourceindex -> nonuniformresource
+
+        auto sb = nonUniformResourceIndexInst->getOperand(0);
+        auto index = nonUniformResourceIndexInst->getOperand(1);
+       
+        auto value = nonUniformResourceIndexInst->getOperand(2);
+        IRBuilder builder(sb);
+        builder.setInsertBefore(nonUniformResourceIndexInst);
+        IRInst* args[] = { sb, index };
+        auto addrInst = builder.emitIntrinsicInst(
+            builder.getPtrType(kIROp_PtrType, value->getFullType(), SpvStorageClassStorageBuffer),
+            kIROp_RWStructuredBufferGetElementPtr,
+            2,
+            args);
+        auto textureInst = builder.emitRWStructuredBufferGetElementPtr(value, index);
+        nonUniformResourceIndexInst->replaceUsesWith(textureInst);
+        nonUniformResourceIndexInst->removeAndDeallocate();      
+        addUsersToWorkList(textureInst);
```

Comment:
Somewhat confused as to what's going on here. I assume this is still WIP?

Reply:
I'm guessing slang would prefer to do this by propagating `NonUniformResource` down to the users, so probably something like the following?
```
auto value = nonUniformResourceIndexInst->GetOperand(0);
traverseUses(nonUniformResourceIndexInst, [&](IRUse* use) 
{
 auto user = use->getUser();
 // Optionally:
 if (user->getOp() != /* ops for which NonUniformResourceIndex makes sense */)
  return;
 IRBuilder builder(user);
 auto decoratedUser = builder.emitNonUniformResourceIndex(user);
 user->replaceUsesWith(decoratedUser);
});
// Optionally also 
// nonUniformResourceIndexInst->replaceUsesWith(value);
```

Reply:
Yes it is still a WIP. I’m hoping to get it ready in a few days. I ll update the title until then. Sorry for the confusion.